x-pack/plugins/lens/public/app_plugin/mounter.tsx

```diff
@@ -270,64 +304,180 @@ export async function mountApp(
   };
 }
 
-export function loadDocument(
+export function loadInitialStore(
   redirectCallback: (savedObjectId?: string) => void,
   initialInput: LensEmbeddableInput | undefined,
   lensServices: LensAppServices,
   lensStore: LensRootStore,
   embeddableEditorIncomingState: EmbeddableEditorState | undefined,
-  dashboardFeatureFlag: DashboardFeatureFlagConfig
+  dashboardFeatureFlag: DashboardFeatureFlagConfig,
+  datasourceMap: Record<string, Datasource>,
+  visualizationMap: Record<string, Visualization>,
+  initialContext?: VisualizeFieldContext
 ) {
   const { attributeService, chrome, notifications, data } = lensServices;
-  const { persistedDoc } = lensStore.getState().app;
+  const { persistedDoc } = lensStore.getState().lens;
   if (
     !initialInput ||
     (attributeService.inputIsRefType(initialInput) &&
       initialInput.savedObjectId === persistedDoc?.savedObjectId)
   ) {
-    return;
+    return initializeDatasources(
+      datasourceMap,
+      lensStore.getState().lens.datasourceStates,
+      undefined,
+      initialContext,
+      {
+        isFullEditor: true,
+      }
+    )
+      .then((result) => {
+        const datasourceStates = Object.entries(result).reduce(
+          (state, [datasourceId, datasourceState]) => ({
+            ...state,
+            [datasourceId]: {
+              ...datasourceState,
+              isLoading: false,
+            },
+          }),
+          {}
+        );
+        lensStore.dispatch(
+          setState({
+            datasourceStates,
+            isLoading: false,
+          })
+        );
+        if (initialContext) {
+          const selectedSuggestion = getVisualizeFieldSuggestions({
```

Comment:
The whole block up to line 384 could be moved into the reduced because it's completely pure. The only thing we would need to make sure is making datasource and visualization map known to the reducer. But this will probably happen in a separate step, right?